### PR TITLE
Fairplay ezdrm

### PR DESCRIPTION
**Pull request by koldo92**
Hi Jhomlala,

I have created a branch or better_player to add support to iOS FairPlay content.

I have seen that FlutterTexture buffering displays black screen with DRM encoded content, so I have change the iOS native plugin to use FlutterPlatformViewFactory with FlutterPlatformView instead of registering textures. I have tried to make the less changes posible with the existing code.

I have tested that this changes doesn't broke existing code on iOS&Android.

Regards,

### CODE REVIEW

**Comment by koldo92** [comment]
This version only works with EZDRM FairPlay key server, because is the one we work with.
I have added a working example using a DRM video and a FairPlay certificate from EZDRM demo project.

**Comment by adrianByv** [comment]
I've been developing with @koldo92 this pull request. We have test it with different hls fairplay encoded videos and works fine.

After you merge the DASH (.mpd) pull request into master we could merge it on this branch. To move certificateUrl to the new DataSource.

File: ios/Classes/DRM/FLTEzdrmAssetsLoaderDelegate.h [hunk 1]
```diff
@@ -0,0 +1,17 @@
+//
```
Comment (jhomlala) [inline ios/Classes/DRM/FLTEzdrmAssetsLoaderDelegate.h:1]: Please remove this comment from the top. I know it's created automatically. There's no similar header comments for other files in Better Player.

File: ios/Classes/DRM/FLTEzdrmAssetsLoaderDelegate.h [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+//  Created by Koldo on 18/05/2021.
+//
+
+#import <Foundation/Foundation.h>
+#import <AVKit/AVKit.h>
+#import <AVFoundation/AVFoundation.h>
+
+
+@interface FLTEzdrmAssetsLoaderDelegate : NSObject
```
Comment (jhomlala) [inline ios/Classes/DRM/FLTEzdrmAssetsLoaderDelegate.h:13]: I think it's good time to change that prefix "FLT". I think this header should be named BetterPlayerEzDRMAssetsLoaderDelegate or BetterPlayerEzDrmAssetsLoaderDelegate.

File: ios/Classes/DRM/FLTEzdrmAssetsLoaderDelegate.m [hunk 3]
```diff
@@ -0,0 +1,113 @@
+//
```
Comment (jhomlala) [inline ios/Classes/DRM/FLTEzdrmAssetsLoaderDelegate.m:1]: Same as above.

File: ios/Classes/DRM/FLTEzdrmAssetsLoaderDelegate.m [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+//  FLTEzdrmAssetsLoaderDelegate.m
+//  better_player
+//
+//  Created by Koldo on 18/05/2021.
+//
+
+#import "FLTEzdrmAssetsLoaderDelegate.h"
+
+@implementation FLTEzdrmAssetsLoaderDelegate
```
Comment (jhomlala) [inline ios/Classes/DRM/FLTEzdrmAssetsLoaderDelegate.m:10]: Same as above please change name.

File: ios/Classes/DRM/FLTEzdrmAssetsLoaderDelegate.m [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+
+NSString * KEY_SERVER_URL = @"https://fps.ezdrm.com/api/licenses/";
+
+- (instancetype)initWithCertificateUrl:(NSURL *)certificateUrl {
+    self = [super init];
+    _certificateURL = certificateUrl;
+    return self;
+}
+
+- (void)removeKeyWith:(NSFileManager *)fileManager {
```
Comment (jhomlala) [inline ios/Classes/DRM/FLTEzdrmAssetsLoaderDelegate.m:22]: Do we need that method? Seems to be empty.

File: ios/Classes/DRM/FLTEzdrmAssetsLoaderDelegate.m [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+
+#import "FLTEzdrmAssetsLoaderDelegate.h"
+
+@implementation FLTEzdrmAssetsLoaderDelegate
+
+NSString *_assetId;
+
+NSString * KEY_SERVER_URL = @"https://fps.ezdrm.com/api/licenses/";
+
+- (instancetype)initWithCertificateUrl:(NSURL *)certificateUrl {
```
Comment (jhomlala) [inline ios/Classes/DRM/FLTEzdrmAssetsLoaderDelegate.m:16]: Can you add comment for this method same as getContentKeyAndLeaseExpiryfromKeyServerModuleWithRequest.

File: ios/Classes/DRM/FLTEzdrmAssetsLoaderDelegate.m [hunk 7]
```diff
@@ -0,0 +1,113 @@
...
+ ** the example here uses a local certificate
+ ** but you may need to edit this function to point to your certificate
+ ** ---------------------------------------*/
+- (NSData *)getAppCertificate:(NSString *) String {
+    NSData * certificate = nil;
+    certificate = [NSData dataWithContentsOfURL:_certificateURL];
+    return certificate;
+}
+
+- (BOOL)resourceLoader:(AVAssetResourceLoader *)resourceLoader shouldWaitForLoadingOfRequestedResource:(AVAssetResourceLoadingRequest *)loadingRequest {
```
Comment (jhomlala) [inline ios/Classes/DRM/FLTEzdrmAssetsLoaderDelegate.m:67]: Please add method comment. Could you format it a little bit?

File: ios/Classes/DRM/FLTEzdrmAssetsLoaderDelegate.m [hunk 8]
```diff
@@ -0,0 +1,113 @@
...
+    NSURL *assetURI = loadingRequest.request.URL;
+    NSString * str = assetURI.absoluteString;
+    NSString * mySubstring = [str substringFromIndex:str.length - 36];
```
Comment (jhomlala) [inline ios/Classes/DRM/FLTEzdrmAssetsLoaderDelegate.m:70]: `str` and `mySubstring` aren't best way to describe what this string does. Can you provide some better names?. Also str.length - 36 - are you sure that it will be always 36?

File: ios/Classes/DRM/FLTEzdrmAssetsLoaderDelegate.m [hunk 9]
```diff
@@ -0,0 +1,113 @@
...
+    _assetId = mySubstring;
+    NSString * scheme = assetURI.scheme;
+    NSData * requestBytes;
+    NSData * certificate;
+    if (!([scheme isEqualToString: @"skd"])){
```
Comment (jhomlala) [inline ios/Classes/DRM/FLTEzdrmAssetsLoaderDelegate.m:75]: Is there any other scheme?

File: ios/Classes/DRM/FLTEzdrmAssetsLoaderDelegate.m [hunk 10]
```diff
@@ -0,0 +1,113 @@
...
+        [dataRequest respondWithData:responseData];
+        [loadingRequest finishLoading];
+    } else {
+        [loadingRequest finishLoadingWithError:error];
+    }
+
+    return YES;
+}
+
+- (BOOL)resourceLoader:(AVAssetResourceLoader *)resourceLoader shouldWaitForRenewalOfRequestedResource:(AVAssetResourceRenewalRequest *)renewalRequest {
```
Comment (jhomlala) [inline ios/Classes/DRM/FLTEzdrmAssetsLoaderDelegate.m:109]: Can you add comment for this method?

File: ios/Classes/FLTBetterPlayer.h [hunk 11]
```diff
@@ -0,0 +1,65 @@
+//
```
Comment (jhomlala) [inline ios/Classes/FLTBetterPlayer.h:1]: Please remove.

File: ios/Classes/FLTBetterPlayer.h [hunk 12]
```diff
@@ -0,0 +1,65 @@
...
+#import <AVFoundation/AVFoundation.h>
+#import <KTVHTTPCache/KTVHTTPCache.h>
+#import <GLKit/GLKit.h>
+#import "FLTTimeUtils.h"
+#import "FLTBetterPlayerView.h"
+#import "FLTEzdrmAssetsLoaderDelegate.h"
+
+NS_ASSUME_NONNULL_BEGIN
+
+@interface FLTBetterPlayer : NSObject <FlutterPlatformView, FlutterStreamHandler, AVPictureInPictureControllerDelegate>
```
Comment (jhomlala) [inline ios/Classes/FLTBetterPlayer.h:20]: Please change name to BetterPlayer. Just remove "FLT".

File: ios/Classes/FLTBetterPlayerPlugin.h [hunk 13]
```diff
@@ -2,7 +2,20 @@
...
 #import <Flutter/Flutter.h>
+#import <AVKit/AVKit.h>
+#import <AVFoundation/AVFoundation.h>
+#import <KTVHTTPCache/KTVHTTPCache.h>
+#import <GLKit/GLKit.h>
+#import "FLTTimeUtils.h"
+#import "FLTBetterPlayer.h"
+#import <MediaPlayer/MediaPlayer.h>
+
+@interface FLTBetterPlayerPlugin : NSObject <FlutterPlugin, FlutterPlatformViewFactory>
```
Comment (jhomlala) [inline ios/Classes/FLTBetterPlayerPlugin.h:15]: Same as above.

File: ios/Classes/FLTBetterPlayerView.h [hunk 14]
```diff
@@ -0,0 +1,9 @@
+#import <UIKit/UIKit.h>
+#import <AVKit/AVKit.h>
+#import <AVFoundation/AVFoundation.h>
+
+// FLTBetterPlayerView.h
+@interface FLTBetterPlayerView : UIView
```
Comment (jhomlala) [inline ios/Classes/FLTBetterPlayerView.h:6]: BetterPlayerView

File: ios/Classes/FLTBetterPlayerView.m [hunk 15]
```diff
@@ -0,0 +1,21 @@
+#import "FLTBetterPlayerView.h"
+
+// FLTBetterPlayerView.m
+@implementation FLTBetterPlayerView
```
Comment (jhomlala) [inline ios/Classes/FLTBetterPlayerView.m:4]: BetterPlayerView

File: ios/Classes/FLTTimeUtils.h [hunk 16]
```diff
@@ -0,0 +1,18 @@
...
+//
+//  Created by Koldo on 17/05/2021.
+//
+
+#import <Foundation/Foundation.h>
+#import <AVKit/AVKit.h>
+
+NS_ASSUME_NONNULL_BEGIN
+
+@interface FLTTimeUtils : NSObject
```
Comment (jhomlala) [inline ios/Classes/FLTTimeUtils.h:13]: BetterPlayerTimeUtils

File: ios/Classes/FLTTimeUtils.h [hunk 17]
```diff
@@ -0,0 +1,18 @@
+//
```
Comment (jhomlala) [inline ios/Classes/FLTTimeUtils.h:1]: Remove

File: ios/Classes/FLTTimeUtils.m [hunk 18]
```diff
@@ -0,0 +1,19 @@
+//
```
Comment (jhomlala) [inline ios/Classes/FLTTimeUtils.m:1]: Remove

File: ios/Classes/FLTTimeUtils.m [hunk 19]
```diff
@@ -0,0 +1,19 @@
...
+//  FLTTimeUtils.m
+//  better_player
+//
+//  Created by Koldo on 17/05/2021.
+//
+
+#import "FLTTimeUtils.h"
+
+@implementation FLTTimeUtils
```
Comment (jhomlala) [inline ios/Classes/FLTTimeUtils.m:10]: BetterPlayerTimeUtils

File: ios/Classes/FLTBetterPlayerView.h (hunk 14 above)
Comment (jhomlala) [reply ios/Classes/FLTBetterPlayerView.h:6]: Also what's the purpose for this class?

File: lib/src/video_player/method_channel_video_player.dart [hunk 20]
```diff
@@ -401,7 +404,15 @@ class MethodChannelVideoPlayer extends VideoPlayerPlatform {
 
   @override
   Widget buildView(int? textureId) {
-    return Texture(textureId: textureId!);
+    if (defaultTargetPlatform == TargetPlatform.iOS) {
+      return UiKitView(
```
Comment (jhomlala) [inline lib/src/video_player/method_channel_video_player.dart:408]: Why did you change Texture to UiKitView. Does UiKitView is better than Texture?

Comment (jhomlala) [reply lib/src/video_player/method_channel_video_player.dart:408]: Does this change won't brake other stuff like PIP mode or notification of the player?

**Comment by jhomlala** [comment]
Guys thank you for this awesome work. I've added my comments to your PR. Please check them in your spare time. Thanks!

File: lib/src/video_player/method_channel_video_player.dart (hunk 20 above)
Comment (adrianByv) [reply lib/src/video_player/method_channel_video_player.dart:408]: Hi. 
> https://github.com/jhomlala/betterplayer/pull/488#discussion_r635503219 :
Why did you change Texture to UiKitView. Does UiKitView is better than Texture?

We migrate a native app using bitmovin to flutter. We start implementing a custom plugin to load bitmovin native sdk in iOS with UiKitView and Android with AndroidView. We detect that iOS part works fine, but with android verions prior to 10.0 it was displaying a black screen (similar to when you try to record the screen on a device playing drm video). Then we check this plugin and looked that working with Texture solved the android problem, but iOS was getting the black screen...
Then we implement a simple flutter plugin with fairlplay and UiKitView and verify that it was working fine on iOS. That's the reason we have change this. Maybe it can be changed to AndroidView too, but we didn't test it. We have more experience with iOS dev...

> https://github.com/jhomlala/betterplayer/pull/488#discussion_r635503537 :
Does this change won't brake other stuff like PIP mode or notification of the player?

We have try to make just the necessary changes to migrate from Texture to UiKitView. We maintain all the process, we just change the way AVPlayer is rendered, generating an UIView based on an AVPlayerLayer that get data from the AVPlayer. PIP process and notifications are related with the AVPlayer and this code is exactly the same. 
Also we have test that it's working fine on real devices (iPhons and iPads) and with production fairplay hls videos.

Maybe this is not the optimum way to implement (the plugin channel can be managed by each View and decentralize it) but it's the best way to maintain all the compatibility for Android and iOS with the fewest changes.

File: ios/Classes/DRM/FLTEzdrmAssetsLoaderDelegate.h (hunk 1 above)
Comment (koldo92) [reply ios/Classes/DRM/FLTEzdrmAssetsLoaderDelegate.h]: Fixed is this commit d03cee3

File: ios/Classes/DRM/FLTEzdrmAssetsLoaderDelegate.m (hunk 3 above)
Comment (koldo92) [reply ios/Classes/DRM/FLTEzdrmAssetsLoaderDelegate.m]: Fixed is this commit d03cee3

File: ios/Classes/DRM/FLTEzdrmAssetsLoaderDelegate.h (hunk 2 above)
Comment (koldo92) [reply ios/Classes/DRM/FLTEzdrmAssetsLoaderDelegate.h]: Fixed is this commit 613901e

File: ios/Classes/DRM/FLTEzdrmAssetsLoaderDelegate.m (hunk 5 above)
Comment (koldo92) [reply ios/Classes/DRM/FLTEzdrmAssetsLoaderDelegate.m]: Fixed in this commit 71a3256

File: ios/Classes/FLTTimeUtils.h (hunk 16 above)
Comment (koldo92) [reply ios/Classes/FLTTimeUtils.h]: Fixed in this commit 2468462

File: ios/Classes/DRM/FLTEzdrmAssetsLoaderDelegate.m (hunk 4 above)
Comment (koldo92) [reply ios/Classes/DRM/FLTEzdrmAssetsLoaderDelegate.m]: Fixed is this commit 613901e

File: ios/Classes/FLTTimeUtils.h (hunk 17 above)
Comment (koldo92) [reply ios/Classes/FLTTimeUtils.h]: Fixed in this commit d03cee3

File: ios/Classes/FLTTimeUtils.m (hunk 18 above)
Comment (koldo92) [reply ios/Classes/FLTTimeUtils.m]: Fixed in this commit d03cee3

File: ios/Classes/FLTTimeUtils.m (hunk 19 above)
Comment (koldo92) [reply ios/Classes/FLTTimeUtils.m]: Fixed in this commit 2468462

**Comment by koldo92** [comment]
You're welcome, thank you for this awesome library :)

**Comment by jhomlala** [comment]
@koldo92 @adrianByv Please let me know if the PR is ready for next check :) Also can you resolve conflict with FLTBetterPlayerPlugin.m?

**Comment by koldo92** [comment]
> @koldo92 @adrianByv Please let me know if the PR is ready for next check :) Also can you resolve conflict with FLTBetterPlayerPlugin.m?

Conflicts resolve on 47361c9

PR is ready!

**Comment by jhomlala** [comment]
Merged. Thank you guys for awesome work!
